stacks/api/bucket/history/watcher.spec.ts

```diff
@@ -0,0 +1,329 @@
+
+    expect(previousSchemaSpy).toHaveBeenCalledTimes(1);
+    expect(previousSchemaSpy).toHaveBeenCalledWith(buckets[0]._id);
+
+    expect(deleteAtPathSpy).toHaveBeenCalledTimes(2);
+    expect(deleteAtPathSpy).toHaveBeenCalledWith(buckets[0]._id, ["title"]);
+    expect(deleteAtPathSpy).toHaveBeenCalledWith(buckets[0]._id, ["description"]);
+  });
+
+  it("should create new history when bucket document is updated", async () => {
+    //update spy as it returns previous bucket document
+    previousDocumentSpy.and.returnValue(Promise.resolve(secondBucketDocuments[1]));
+
+    let updatedDocument = {...secondBucketDocuments[1], name: "updated second name", age: 21};
+
+    await module
+      .get(DatabaseService)
+      .collection(`bucket_${buckets[1]._id}`)
+      .replaceOne({_id: updatedDocument._id}, updatedDocument);
+
+    //wait until changes detected
+    await new Promise(resolve => setTimeout(resolve, 100));
+
+    expect(previousDocumentSpy).toHaveBeenCalledTimes(1);
+    expect(previousDocumentSpy).toHaveBeenCalledWith(buckets[1]._id, secondBucketDocuments[1]._id);
+
+    expect(insertOneSpy).toHaveBeenCalledTimes(1);
+    expect(insertOneSpy).toHaveBeenCalledWith({
+      bucket_id: buckets[1]._id,
+      document_id: secondBucketDocuments[1]._id,
+      changes: [
+        {
+          kind: 1,
+          path: ["name"],
+          patches: [
+            {diffs: [[-1, "updated "], [0, "seco"]], start1: 0, start2: 0, length1: 12, length2: 4}
+          ]
+        },
+        {
+          kind: 1,
+          path: ["age"],
+          patches: [
+            {diffs: [[0, "2"], [-1, "1"], [1, "0"]], start1: 0, start2: 0, length1: 2, length2: 2}
+          ]
+        }
+      ]
+    });
+  });
+
+  it("should throw error when there is no changes between previous and updated documents", async done => {
+    //provide only unhandledRejection listener on process is our custom listener then revert changes back
+    const listeners = process.listeners("unhandledRejection");
+    process.removeAllListeners("unhandledRejection");
+    const handleError: NodeJS.UnhandledRejectionListener = (reason, promise) => {
+      expect(reason instanceof Error).toBe(true);
+      expect(reason["message"]).toBe(
+        "Database propagated changes but state of previous and current documents is equal so there is no change in between documents. This usually happens when there is no replication member that has slaveDelay."
+      );
+      process.removeListener("unhandledRejection", handleError);
+      listeners.forEach(l => process.addListener("unhandledRejection", l));
+      done();
+    };
+    process.addListener("unhandledRejection", handleError);
+
+    //update spy as it returns same bucket document
+    const updatedDocument = {...firstBucketDocuments[0], title: "new title"};
+    previousDocumentSpy.and.returnValue(Promise.resolve(updatedDocument));
+
+    await module
+      .get(DatabaseService)
```

Comment:
Please put database service into a variable to reuse. 
It would be easier to read.

Reply:
Please check my last commit.